src/expected_value.py

```diff
@@ -0,0 +1,95 @@
+                sell_token = swap.pair.token0.symbol
+                buy_token = swap.pair.token1.symbol
+                sell_amount = swap.amount0_in
+                buy_amount = swap.amount1_out
+                volume = swap.amount_usd
+            else:
+                sell_token = swap.pair.token1.symbol
+                buy_token = swap.pair.token0.symbol
+                sell_amount = swap.amount1_in
+                buy_amount = swap.amount0_out
+                volume = swap.amount_usd
+            o = {
+                'sellToken': sell_token,
+                'buyToken': buy_token,
+                'sellAmount': sell_amount,
+                'buyAmount': buy_amount,
+                'volume': volume
+            }
+            block_number = int(swap_transaction.block_number)
+            if block_number not in swaps_by_block.keys():
+                swaps_by_block[block_number] = []
+            swaps_by_block[block_number].append(o)
+    return swaps_by_block
+
+if not use_cache:
+    swaps_by_block = get_uniswap_swaps()
+    #sorted(swaps_by_block.keys(), reverse=True)
+    with open("uniswap_swaps.pickled", "bw+") as f:
+        pickle.dump(sorted(swaps_by_block.keys(), reverse=True), f)
+else:
+    with open("uniswap_swaps.pickled", "br") as f:
+        swaps_by_block = pickle.load(f)
+
+#print(swaps_by_block)
+#print({frozenset([o['sellToken'], o['buyToken']]) for b in swaps_by_block.values() for o in b})
+
+print("Expected waiting time for opposite offer in block:")
+number_of_blocks_looking_forward = 50
+focus_pairs=[['WETH', 'USDT', 'DAI', 'USDC'],['WETH', 'DAI'], ['WETH', 'USDT'],['WETH', 'USDC'] , ['WETH', 'MKR'], ['WETH', 'GNO'], ['WETH', 'SUSHI'], ['WETH', 'SWRV'], ['WETH', 'WBTC'], ['WETH', 'YFI'], ['WETH', 'LINK'] ]
+sorted_blocks = sorted(swaps_by_block.keys(), reverse=True)
+for focus_pair in focus_pairs:
+    prob_opposite_offer = dict()
+    for k in range(number_of_blocks_looking_forward):
+        nr_blocks_with_at_least_one_opposite_offer =0
+        avg_volume = 0
+        nr_direct_trades = 0
+        for i in range(len(sorted_blocks) - k):
+            block_i = sorted_blocks[i]
+            buy_tokens_i = {o['buyToken'] for o in swaps_by_block.get(block_i, [])}
+            sell_tokens_j = {
+                o['sellToken']
+                for j in range(block_i, block_i + k + 1)
+                for o in swaps_by_block.get(j, [])
+            }
+            if focus_pair is not None:
+                if focus_pair[0] in buy_tokens_i and \
+                    (focus_pair[1] in sell_tokens_j or \
+                    (len(focus_pair) >= 3 and focus_pair[2] in sell_tokens_j)  or \
+                    (len(focus_pair) >= 4 and focus_pair[3] in sell_tokens_j)
+                    ):
+                    nr_blocks_with_at_least_one_opposite_offer += 1
+                
+        prob_opposite_offer[k] = nr_blocks_with_at_least_one_opposite_offer / (len(sorted_blocks) - k)
```

Comment:
If I understand correctly, what is being computed here is:
```
p(sell t1 for t2 on a range of blocks with size k)
```
Right?

Since this changes due to the frequency of the token_pair trades, I wonder if the conditional is also interesting:
```
p(sell t1 for t2 on a range of blocks with size k | t1 for t2 is sold on the exchange)
=
nr_blocks_with_at_least_one_opposite_offer / nr_blocks_with_at_least_one_opposite_offer_over_all_blocks
```

Which would give a "normalized" view over all token pairs (hence should not change much between token pairs).

Reply:
What I was trying to explain is the following:

O - order
C - counter order

Say we have 1000 blocks, 
``` 
1 2 3 4 5 ... 1000
O C O   C
```
If I understood the code correctly, this line
```
prob_opposite_offer[k] = nr_blocks_with_at_least_one_opposite_offer / (len(sorted_blocks) - k)
```
computes the probability that an order and a counter-order are at most k-1 blocks apart. So that would be 1/1000 for k=2 our example above.

What I'm saying is that doesn't reflect the "popularity" of the pair. I think that the probability of being matched in the next block should be 0.5 - because the two times someone posted an order, the counter-order appeared in the next block half of the times.

I hope that makes it clear :)

Reply:
> So that would be 1/1000 for k=2 our example above.

No, it computes the probability that one additional order would find a match, a counter order C. Hence it only looks into the counterorders. For k=2, it would be the slots #{1,2,3,4,5,}=5/1000

Is this logic screwed? I am not soo good with it.

Reply:
I believe that with k=2 we would increment `nr_blocks_with_at_least_one_opposite_offer` only once, when i=0, and therefore
```
nr_blocks_with_at_least_one_opposite_offer = 1
```
.

But perhaps that is not relevant. My whole point is that the size of the sample we're looking at (1000) shouldn't matter to compute what is essential a conditional probability: the probability that a counter-order for o is found within k blocks from now *given that* I posted my order o now. 

We could look only at "counter-orders", as you're doing, but for that I believe we need to assume that the appearance of orders and counter-orders is independent. And I'd argue that it should be not independent at all because of arbitrage - once a trader posts and order, there will likely be adjustments following it by arbitrageurs. This should be more noticeable in low volume trading pairs.

Reply:
To make it formal:

We want to compute:
```
p(counter_order_in_next_k_blocks | order_now) = p(counter_order_in_next_k_blocks, order_now) / p(order_now)
```

and only if you assume that the events `counter_order_in_next_k_blocks` and `order_now are independent`, you can do:
```
p(counter_order_in_next_k_blocks | order_now) = p(counter_order_in_next_k_blocks)
```

Reply:
Sorry, I must have missed that thread. Yes, I see and agree.

Reply:
I added a comment explaining this fact.